subsys/event_manager_proxy/event_manager_proxy.c

```diff
@@ -0,0 +1,490 @@
+			return;
+		}
+	}
+
+	k_event_set(&emp_all_remotes_started, 0x1);
+}
+
+static void handle_remote_command(struct emp_ipc_data *ipc, const void *data, size_t len)
+{
+	const struct emp_cmd *cmd = data;
+
+	if (len < sizeof(*cmd)) {
+		LOG_ERR("Unexpected command size: %zu", len);
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	switch (cmd->code) {
+	case EMP_CMD_REGISTER:
+		handle_remote_command_register(ipc, data, len);
+		break;
+
+	case EMP_CMD_START:
+		handle_remote_command_start(ipc, data, len);
+		break;
+
+	default:
+		LOG_ERR("Unsupported command %u", cmd->code);
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+}
+
+/**
+ * @brief The data received on endpoint callback
+ *
+ * This callback is called when there is some data received from the remote core.
+ *
+ * @param data The pointer to the data received.
+ * @param len  The length of the data received.
+ * @param priv The pointer of the related element of the @ref emp_ipc_data array.
+ */
+static void handle_ipc_data_receive(const void *data, size_t len, void *priv)
+{
+	struct emp_ipc_data *ipc = priv;
+
+	/* Execute only from threads! */
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	if (ipc->started && emp_started) {
+		handle_remote_event(ipc, data, len);
+	} else {
+		handle_remote_command(ipc, data, len);
+	}
+}
+
+/**
+ * @brief The error on the endpoint callback
+ *
+ * @param message The message from the backend.
+ * @param priv    The pointer of the related element of the @ref emp_ipc_data array.
+ */
+static void handle_ipc_endpoint_error(const char *message, void *priv)
+{
+	LOG_ERR("Endpoint error: \"%s\"", log_strdup(message));
+	__ASSERT_NO_MSG(false);
+}
+
+static int send_event_to_remote(struct emp_ipc_data *ipc, const struct event_header *eh)
+{
+	const struct event_type *remote_ev = ipc->event_type_map[et2idx(eh->type_id)];
+
+	if (remote_ev == NULL) {
+		return 0;
+	}
+
+	size_t size = event_manager_event_size(eh);
+	uint32_t buffer[ceiling_fraction(size, sizeof(uint32_t))];
+	struct event_header *remote_eh = (struct event_header *)buffer;
+
+	memcpy(buffer, eh, sizeof(buffer));
+	remote_eh->type_id = remote_ev;
+
+	int ret = ipc_service_send(&ipc->ept, buffer, sizeof(buffer));
+
+	if (ret < 0) {
+		LOG_ERR("Cannot send event to remote %p", ipc);
+		__ASSERT_NO_MSG(false);
+	}
+
+	return ret;
+}
+
+static void event_manager_proxy_on_event_process(const struct event_header *eh)
+{
+	int ret = 0;
+
+	if (!emp_started) {
+		return;
+	}
+
+	for (size_t i = 0; (i < ARRAY_SIZE(emp_ipc_data)) && !ret; ++i) {
+		struct emp_ipc_data *ipc = &emp_ipc_data[i];
+
+		if (!ipc->used || !ipc->started) {
+			continue;
+		}
+
+		ret = send_event_to_remote(ipc, eh);
+	}
+}
+EVENT_HOOK_POSTPROCESS_REGISTER(event_manager_proxy_on_event_process);
+
+static int add_ipc_instace(struct emp_ipc_data *ipc, const struct device *instance)
+{
+	int ret = ipc_service_open_instance(instance);
+
+	if (ret && ret != -EALREADY) {
+		LOG_ERR("IPC service open instance failure: %d", ret);
+		return ret;
+	}
```

Comment:
But we do - we go here only if ret is not 0 and not -EALREADY. So for -EALREADY the code continues.

Reply:
my bad. For some reason I hooked that the emp is already set up but indeed this is only about the ip service instance.
But one thing bothers me in such case - if we are the owner and sole user shouldn't the instance be uninitialized?

Reply:
Other module may already initialize the same instance and communicate with it using different endpoins. It is no the case in icmsg backend - as there is only one endpoint allowed. But using OpenAMP gives us such a possibility. But I can see another issue here - if event_manager_proxy_add_remote is called multiple times with the same instance - it would be added as a two different entries. NCSDK-14665 created to cover the case.